### PR TITLE
PDEP-1 Revision (Decision Making)

### PR DESCRIPTION
**Pull request by noatamir**
The governance working group has been discussing the PDEP workflow and decision-making for some time now. **Feedback is welcome.**

This PR proposes some formulation of the existing workflow to clarify and facilitate it for current and future authors, and voting members.
In our discussions, we assessed that a more structured process would help folks who submit the PDEP know how and when it can progress, and how to better engage with others.

The timelines were considered to facilitate an engaging discussion, and automation. We aim to make the lives of the PDEP authors easier by taking away the need to decide "has it been long enough", or "cat herding" folks to engage and clarify their comments so that PDEPs don't get large change requests that could have been clarified early on - late in the process.

The quorum was discussed a lot in our working group, and we're unsure about the best strategy here. We were looking for a sufficient number so if a decision is passed one feels it's valid (e.g. maybe 2 people is too low), but not too high so that we feel like ongoing discussions are blocked because it's too hard to get that many voting members involved. Further proposal on this point are welcome.

We tried to make things as flexible and clear as possible. Being mindful of the needs of the authors, as well as those engaging in the discussion, be them voting members or not.

At this point, we would like to invite feedback. It might be that you need more clarity on the suggestions, have ideas for improvement, or aren't convinced that this will affect the current process positively.

Co-authored-by: @MarcoGorelli, @rhshadrach, @Dr-Irv, @jorisvandenbossche

### CODE REVIEW

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (Dr-Irv) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:244]:
```suggestion
- 09 June 2023: Version 3([GH-53576][53576]) defines a structured decision making process for PDEPs
```

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
some small things

**Review by WillAyd — APPROVED** [review]
very nice - lgtm

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (phofl) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:102]: Can we add ``close the discussion period preemptively`` here?

**Comment by topper-123** [comment]
+1 from me.

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (noatamir) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:102]:
```suggestion
- In case 15 days passed without any new comments, the authors may close the discussion period preemptively, 
```

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (phofl) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md:132]: Scikit learn seems to require a simple majority

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (attack68) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:111]: Apologies if I havce missed this but it does appear that a "voting member" is explicitly defined?

The removed section seems to suggest that a voting member is anyone in the "core development team". Does pandas have any segregation between "active" and "inactive" core development members and does this have any bearing on  their right to vote?

Is there any scope to specifically invite certain informed (non-core dev) parties to vote on issues? I would probably consider this an abuse of power and instead invite to participate in the discussion to influence members.

If votes are cast by core dev members does this have implications for inviting future members into the core dev team? i.e to not push for adding members of a group of similar, or friendly like minded individuals, perhaps from the same institution etc, that can build up substantial iunfluence?

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (attack68) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:118]: Would it not be sufficient for a voting member to follow a discussion led by knowledgeable parties and simply provide a vote that read: "-1: disapprove. I agree with XXX's analysis and concerns and disagree that YYY's argument and the PDEP are suitable".

Rather than provide rules that restrict voting in certain ways (which might be difficult to police anyway), would it not be sufficient to design quorum and majority rules that aim to progress PDEPs that have garnered 'sufficient' support (whatever that may end up being), above objections.

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (attack68) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:110]: If the VOTE issues are public (with anyone allowed to post) it might be worth commenting that VOTE issues will state their intention to delete any vote that is not valid, i.e. if it is posted by a non-voting member?

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (attack68) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:139]: Should there be an explicit statement discussing what happens if a quorum is not reached, i.e. if not enough voting members participate in the vote?
Does that mean the proposal is explicitly rejected and cannot be resubmitted?
Or does it mean that the PDEP may undergo a period of hiberation followed by a new period of discussion and a new vote is permitted after at least (4?) months of the previous vote?

Comment (phofl) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md:139]: Why shouldn't it be possible to resubmit pdeps if circumstances change? How would decide if it's the same pdep or something different if some parameters are changed? I'd rather not worry about this now. Most PDEPs will be submitted by core-devs, and I think there isn't anyone in the group who wants to spam us with proposals

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (Dr-Irv) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md:114]: That's correct.

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (Dr-Irv) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md:110]: I think that's a good addition

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (Dr-Irv) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md:139]: > Should there be an explicit statement discussing what happens if a quorum is not reached, i.e. if not enough voting members participate in the vote?

We haven't discussed this possibility.

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (attack68) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md:114]: Could we add this to the doc to make it explicit, and also could we add to the tallies so that the votes are more transparent, i.e. have Approved-Abstained-Disapproved-NoResponse.

It may be relevant for audit at some point in the future to see who / how many people were eligible to vote.

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (noatamir) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md:139]: We discussed this in our governance meeting today. We propose that not reaching quorum is equivalent to rejecting because effectively the PDEP was not accepted in time. The authors may decide to resubmit any rejected PDEP, as long as they add a link to the previous discussion in the new PDEP proposal. As before, we expect PDEP authors to use their own judgement when they decide to resubmit a PDEP proposal (or withdraw one), as to the ongoing relevance and likelihood of reaching quorum and/or acceptance for the resubmission. No hibernation period is explicitly required.
I'll add a change to reflect this soon.

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (noatamir) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md:132]: We discussed this in our governance meeting today, and at least within the group today, we all found the 75% appropriate. 

The 75% came from a consensus seeking decision-making approach, see [this wiki](https://en.wikipedia.org/wiki/Consensus_decision-making#Decision_rules).
The idea was that we don't to have a decision which had many negative votes. 

Personally, I also like the "unanimity minus three" approach - so that once 4 people are against something, the group shouldn't accept it. But I think it needs to appeal to everyone, so it would be great to get more opinions.

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (lithomas1) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:129]: IDK if this has been asked before, but can you vote for your own PDEP?

(Getting a quorum might be more challenging for PDEPs with e.g. 2/3 authors)

Comment (noatamir) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md:129]: I assumed the authors would vote - yes. Happy to make that explicit.

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (jorisvandenbossche) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md:136]: We also though it could be helpful to make the 75% number more tangible with a concrete example: if 12 people vote, 4 negative votes would block it; if 16 people vote, 5 negative votes would block it.

(and so this current proposal is less strict than a "unanimity minus three" rule)

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (jorisvandenbossche) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:114]: Do we want to add a small explanation of how we understand "abstaining" or what it could mean? 
Something like "Not fully convinced, but don't want to block it", or better worded (although there might be many reasons to vote +0, like "I haven't closely followed, but trust the majority")

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @noatamir generally lgtm. Should probably have a section on voting members even if for now is a one-liner... active members of the core team. (active/inactive has been defined/agreed elsewhere?)

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (simonjayhawkins) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md]: > Right now, it is the core team, but it is likely we will not use that in the future.

IMO we should also require a discussion to change this, hence my comment to define voting member in this PDEP now.

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (noatamir) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:102]:
```suggestion
- In case 15 days passed without any new unaddressed comments, the authors may close the discussion period
```

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (noatamir) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:111]:
```suggestion
Each voting member, including author(s) with voting rights, may cast a vote by adding one of the following comments:
```

**Review by simonjayhawkins — APPROVED** [review]

**Review by mroeschke — APPROVED** [review]

**Review by MarcoGorelli — APPROVED** [review]
let's ship it

**Review by attack68 — APPROVED** [review]

**Review by bashtage — APPROVED** [review]
LGMT.

**Review by lithomas1 — APPROVED** [review]

**Review by twoertwein — APPROVED** [review]

**Review by rhshadrach — APPROVED** [review]

**Review by lukemanley — APPROVED** [review]

**Comment by noatamir** [comment]
Curveball: I have been thinking about the voting majority a lot, and I am leaning toward making it a simple majority or 60% at most (e.g. 7/11 approving votes). 
My thinking is that I would like to see interesting and important discussions that help shape the outcome, but I don't want decisions to be hard, actually. 
My caution is that if many decisions pass this way, we might need to re-evaluate because it can alienate anyone who is consistently feeling like they have no say. But, I hope that the discussion period will, in most cases, gives time for authors to reconsider their proposals based on discussion - even if they "have enough votes". 
Do folks want to discuss this further in the next governance meeting?

**Comment by phofl** [comment]
I agree with @noatamir, 75% is too high for me to feel comfortable

**Comment by attack68** [comment]
I disagree. I recall the arguments made in the original governance meeting. With 75%, a 3:1 vote passes and a PDEP is accepted. With 60% 2 in 5 people can oppose and a PDEP still be accepted. For a project which is expected to have a very clear and defined development structure a 3:2 vote is too divided and should not be accepted, IMO.

**Comment by simonjayhawkins** [comment]
> My caution is that if many decisions pass this way, we might need to re-evaluate because it can alienate anyone who is consistently feeling like they have no say.

I think the opposite could be true.

From a recent personal perspective, I felt that the support for the pyarrow dependency was likely to be so strong that I was happy not to challenge the proposal and only if it failed to get through would get involved to help shape the proposal to something acceptable to all.

In order to downvote I was 'required' to participate. And to be fair, the proposal was amended following my comments and additional discussion and I am now more comfortable with it.

But OTH, I was also happy to proceed with a vote as there was maybe a chance that even with 75% there was a chance it would have been accepted and therefore these amendments and further discussion has potentially slowed the process.

So I would argue the opposite. We need a higher threshold so that people feel they have a say.

But I appreciate that if we go much higher, we could effectively return to the veto system.

**Comment by simonjayhawkins** [comment]
> With 75%, a 3:1 vote passes and a PDEP is accepted. With 60% 2 in 5 people can oppose and a PDEP still be accepted. For a project which is expected to have a very clear and defined development structure a 3:2 vote is too divided and should not be accepted, IMO.

4:1 is 80%, 5:1 is 83.33% and 6:1 is 85.7%

5:1 means that if 2 people oppose a proposal, we need at least 12 members to vote or change the proposal in order to be accepted.

So 5:1 could lead to difficult discussions/slow progress, but increasing to 80% (4:1) does not seem unreasonable to me.

**Comment by phofl** [comment]
While I agree that 3-2 is not a great result, I don't think that this is a realistic outcome. We require at least 11 votes for a PDEP to possibly get accepted. I am not comfortable with a 8-3 result not being sufficient to pass a PDEP, same for 11-4. Personally, I feel that both of these results should be strong enough majorities to accept a PDEP

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (phofl) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:102]: I have a question about this: Is it required to give notice that the voting will start in x days or can you just say voting starts now? This is a bit confusing if that happens very early because discussion did not gain traction.

Comment (Dr-Irv) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md:102]: The concept is that you can say "voting starts now"

**Comment by Dr-Irv** [comment]
> While I agree that 3-2 is not a great result, I don't think that this is a realistic outcome. We require at least 11 votes for a PDEP to possibly get accepted. I am not comfortable with a 8-3 result not being sufficient to pass a PDEP, same for 11-4. Personally, I feel that both of these results should be strong enough majorities to accept a PDEP

I agree.  That's why I felt good about the 75% rule.  Going above that seemed to high, below seemed to low.  Our current policy of 100% isn't sustainable.

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (phofl) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md:102]: Can we require at least a discussion period of 3 weeks? I was a bit surprised that this was already ready. Granted, it was my fault to not object to the 75% majority earlier (or follow up on my comment above), but I expected that I'd have more time.

My reasoning here is: Get a earliest moment in time when the voting procedure could possibly start

**Comment by phofl** [comment]
> > While I agree that 3-2 is not a great result, I don't think that this is a realistic outcome. We require at least 11 votes for a PDEP to possibly get accepted. I am not comfortable with a 8-3 result not being sufficient to pass a PDEP, same for 11-4. Personally, I feel that both of these results should be strong enough majorities to accept a PDEP
> 
> I agree. That's why I felt good about the 75% rule. Going above that seemed to high, below seemed to low. Our current policy of 100% isn't sustainable.

But 75% would not be sufficient to pass a 8-3 or 11-4, which feels off to me

**Comment by Dr-Irv** [comment]
> But 75% would not be sufficient to pass a 8-3 or 11-4, which feels off to me

Maybe we lower it to 70% ?

**Comment by MarcoGorelli** [comment]
TBH I'd be OK with 11-4 not passing. Perhaps we can go ahead with what's been agreed upon, and then revise if there are many close calls like that?

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (Dr-Irv) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md:102]: Well, the way it is written is that if there is no discussion for 15 days, you could call for a vote.  So in theory, if there was no discussion for 15 days right after a PDEP was created, the author could call for a vote, but I think that is highly unlikely.

Or we can say that there is a minimum discussion period of 3 weeks, as you suggest.

Comment (phofl) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md:102]: I'd like the 3 weeks, it seems like a low requirement as you said and would make things clearer. I previously assumed that it would require at least 45 days of discussion (that was on me though).

**Comment by lithomas1** [comment]
Since there is already a higher barrier to disapproving(you have to engage in the PDEP discussion first), I think the current limit is fine.

We can always revisit this if necessary, in the future.

**Comment by noatamir** [comment]
Thanks for the responses, folks!
I was expecting a higher number of votes in most cases, considering the quorum requirement.  I don't anticipate that we will have many cases of 8 abstains and a 2-1 split.
Likewise, I agree that we can see if we encounter frustrations and make changes in either approach. 

One last question, is anyone apposed to me adding a requirement of minimum discussion time? I was thinking that 3–4 weeks would be the minimum before an early vote can be triggered. This was intended to support the lower majority threshold, but I think it fits in any case.

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (noatamir) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:102]:
```suggestion
- After 45 days, in case 15 days passed without any new unaddressed comments, the authors may close the discussion period
```

Comment (simonjayhawkins) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md:102]: Since there is a requirement to participate in order to downvote, then we could have a situation where a discussion has not had any objections and proceeding to a vote would be pointless since only those in support could vote.

This is maybe not a bad thing as it puts the onus on members to get involved in the discussion early.

Extending the minimum discussion period is a potential solution to give sufficient opportunity to participate, but it maybe that we should have a notice period for the vote to begin. "voting starts now" is a bit final.

Comment (jorisvandenbossche) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md:102]: > Can we require at least a discussion period of 3 weeks? I was a bit surprised that this was already ready. 

FWIW, this PR wasn't necessarily "ready", it's just that my question if there were more comments triggered people to start approving .. 

But +1 on requiring at least 3 weeks (which already had passed in this case, though)

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (Dr-Irv) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:123]:
```suggestion
where w stands for the total of approving, x of abstaining, y of disapproving votes cast, and z
```

**Comment by Dr-Irv** [comment]
@noatamir what do we need to close this out?

**Comment by simonjayhawkins** [comment]
probably controversial, but closing as stale!

**Comment by phofl** [comment]
Yeah you are correct, I am -1 on closing

**Comment by simonjayhawkins** [comment]
almost 6 months without a response to what is needed to close this out.

@phofl can you answer that?

**Review by simonjayhawkins — DISMISSED** [review]
It seems there was an unaddressed comment. https://github.com/pandas-dev/pandas/pull/53576#discussion_r1478418468

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (simonjayhawkins) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md:102]: to avoid this then maybe we should just remove the requirement to participate in the discussion in order to be able to downvote

**Comment by jorisvandenbossche** [comment]
As far as I remember, the main point of contention last summer was the exact threshold to require for acceptance. The text currently says "a majority of 75% of the non-abstaining votes", but there were some in favor of lowering this threshold. 

I think the main relevant discussion on this PR starts at https://github.com/pandas-dev/pandas/pull/53576#issuecomment-1631555955, continuing in the comments below.

I started a poll with the core team to get an idea of what threshold would be most acceptable.

**Comment by simonjayhawkins** [comment]
Thanks @jorisvandenbossche 

Great to know the process hasn't been abandoned.

I think that the PDEP process (which is ironically this PR) should probably include processes for the scenarios where the PDEP process doesn't follow the suggested timescales and appears abandoned. This also applies to VOTE issues that is not closed after the voting period ends. (for instance do votes still count until the issue is closed?)

A PDEP or PDEP revision that is not explicitly rejected is also not accepted.

I think we need to be able to close stale PDEP PRs or voting issues rather than leaving open with zero communication so that it is clearer to the community what the status is.

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (Dr-Irv) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:102]:
```suggestion
To enable and encourage discussions on PDEPs, we follow a notification schedule. At each of the
following steps, the pandas team, and the pandas dev mailing list are notified via GitHub and
E-mail, with an indication that a voting period will commence afterr15 days of the notification date if there is no further discussion:
- Once a PDEP is ready for discussion.
- After 30 discussion days, with 30 days remaining for discussion.
- After 45 discussion days, with 15 days remaining for discussion.
```

Comment (jorisvandenbossche) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md]: > Extending the minimum discussion period is a potential solution to give sufficient opportunity to participate, but it maybe that we should have a notice period for the vote to begin. "voting starts now" is a bit final.

It's a good point that when a discussion has been quiet, being able to just "start the voting" with an early vote, sounds a bit abrupt (exactly because the discussion went quiet, people might have forgotten about commenting on the PDEP).

I updated the text around this slightly to just state in general that we expect a notification 15 days before a vote would start. And then this heads up for a nearing vote can either be after 30 days for an early vote (so the end result is the same as Noa's suggestion above that the actual cote can happen after 45 days) or later.

Comment (jorisvandenbossche) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md]: @Dr-Irv I pushed a different commit with an attempt to address the same thing, see https://github.com/pandas-dev/pandas/pull/53576/commits/88a7db99fef424bc6614d4ba536a6b3587ff7140. Can you take a look at that?

Comment (Dr-Irv) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md]: @jorisvandenbossche  I looked at just that commit and made a comment on that text.

Comment (simonjayhawkins) [reply web/pandas/pdeps/0001-purpose-and-guidelines.md]: Thanks @jorisvandenbossche 

We use the term "unaddressed comment" in the proposal. I changed my review here to "requires changes" recently because of this.

Now, IIRC in the discussion about PDEP-8 (i've not checked the timelines thoroughly and it maybe that I should use a hypothetical case instead) that we had potentially progressed to the vote with unaddressed comments.

Any ideas to cover this scenario or whether we need to at this stage?

**Review by Dr-Irv — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
@pandas-dev/pandas-core please take a final look / provide final comments, or otherwise (re-)approve the PR!

Main changes since last discussion round:
- Changed the threshold from 75% to 70%
- In the discussion timeline, tweaked the language around an early vote if there was no more discussion for 15 days (after a minimum of 30 days)

**Review by alimcmaster1 — APPROVED** [review]

**Review by mroeschke — APPROVED** [review]

**Review by MarcoGorelli — APPROVED** [review]
no objections

**Review by WillAyd — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
One more small change: I added an extra "Withdrawn" status in addition to "Rejected", for the case the author(s) themselves decide to not get to a vote if they are convinced from the discussion it's eventually not a good idea or when there is an alternative proposal that is decided to push forward, but we still want to merge the PDEP to document the discussion (this status is also used in PEPs, and also numpy and other projects).

**Review by lukemanley — APPROVED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
A final reminder here, I plan to merge this by Friday, assuming we have consensus on those changes.

**Review by phofl — APPROVED** [review]

**Review by TomAugspurger — APPROVED** [review]

**Review by rhshadrach — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks all. Seems like we have consensus here.

**Comment by jorisvandenbossche** [comment]
And thanks @noatamir and the initial write-up, and everyone for the feedback!

**Pull request merged**
